File: backend/pkg/api/datasource/handlers.go
package datasource

import (
	"backend/pkg/api/middleware"
	"backend/pkg/api/response"
	"backend/pkg/api/validation"
	"backend/pkg/core"
	coreDS "backend/pkg/core/datasource"
	"backend/pkg/types"

	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"
)

// ListDataSources handles GET /data-sources
// @Summary List data sources
// @Description Get a list of all data sources for the authenticated user's organization
// @Tags data-sources
// @Accept json
// @Produce json
// @Param page query int false "Page number" default(1)
// @Param limit query int false "Items per page" default(20)
// @Success 200 {object} response.SuccessResponse{data=[]types.DataSourceResponse}
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /data-sources [get]
func ListDataSources(c *fiber.Ctx) error {
	orgID := middleware.GetOrgID(c)

	if orgID != nil {
		sources, err := coreDS.List(c.Context(), *orgID)
		if err != nil {
			return response.Error(c, err)
		}
		return response.Success(c, sources)
	}

	// If no org, return all with pagination
	page := c.QueryInt("page", 1)
	limit := c.QueryInt("limit", 20)

	sources, total, err := coreDS.ListAll(c.Context(), page, limit)
	if err != nil {
		return response.Error(c, err)
	}

	return response.Success(c, fiber.Map{
		"data_sources": sources,
		"total":        total,
		"page":         page,
		"limit":        limit,
	})
}

// GetDataSource handles GET /data-sources/:id
// @Summary Get data source
// @Description Get a data source by ID
// @Tags data-sources
// @Accept json
// @Produce json
// @Param id path string true "Data Source ID"
// @Success 200 {object} response.SuccessResponse{data=types.DataSourceResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /data-sources/{id} [get]
func GetDataSource(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid data source ID"))
	}

	ds, err := coreDS.GetByID(c.Context(), id)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, ds)
}

// CreateDataSource handles POST /data-sources
// @Summary Create data source
// @Description Create a new data source (preferred website to crawl)
// @Tags data-sources
// @Accept json
// @Produce json
// @Param request body types.DataSourceCreateRequest true "Data source details"
// @Success 201 {object} response.SuccessResponse{data=types.DataSourceResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 401 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /data-sources [post]
func CreateDataSource(c *fiber.Ctx) error {
	var req types.DataSourceCreateRequest
	if err := c.BodyParser(&req); err != nil {
		return response.Error(c, core.InvalidInputError("Invalid request body"))
	}
	if err := validation.ValidateStruct(req); err != nil {
		return response.Error(c, err)
	}

	orgID := middleware.GetOrgID(c)

	ds, err := coreDS.Create(c.Context(), orgID, req)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Created(c, ds)
}

// UpdateDataSource handles PUT /data-sources/:id
// @Summary Update data source
// @Description Update an existing data source
// @Tags data-sources
// @Accept json
// @Produce json
// @Param id path string true "Data Source ID"
// @Param request body types.DataSourceUpdateRequest true "Data source update details"
// @Success 200 {object} response.SuccessResponse{data=types.DataSourceResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /data-sources/{id} [put]
func UpdateDataSource(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid data source ID"))
	}

	var req types.DataSourceUpdateRequest
	if err := c.BodyParser(&req); err != nil {
		return response.Error(c, core.InvalidInputError("Invalid request body"))
	}

	ds, err := coreDS.Update(c.Context(), id, req)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, ds)
}

// DeleteDataSource handles DELETE /data-sources/:id
// @Summary Delete data source
// @Description Delete a data source by ID
// @Tags data-sources
// @Accept json
// @Produce json
// @Param id path string true "Data Source ID"
// @Success 200 {object} response.SuccessResponse{data=object{success=boolean}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /data-sources/{id} [delete]
func DeleteDataSource(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid data source ID"))
	}

	if err := coreDS.Delete(c.Context(), id); err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, fiber.Map{"success": true})
}

// TriggerCrawl handles POST /data-sources/:id/crawl
// @Summary Trigger crawl
// @Description Trigger a manual crawl for a data source
// @Tags data-sources
// @Accept json
// @Produce json
// @Param id path string true "Data Source ID"
// @Success 200 {object} response.SuccessResponse{data=object{success=boolean,message=string}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /data-sources/{id}/crawl [post]
func TriggerCrawl(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid data source ID"))
	}

	if err := coreDS.TriggerCrawl(c.Context(), id); err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, fiber.Map{
		"success": true,
		"message": "Crawl triggered successfully",
	})
}

// GetDataSourceContent handles GET /data-sources/:id/content
// @Summary Get data source content
// @Description Get crawled content for a data source
// @Tags data-sources
// @Accept json
// @Produce json
// @Param id path string true "Data Source ID"
// @Param page query int false "Page number" default(1)
// @Param limit query int false "Items per page" default(20)
// @Success 200 {object} response.SuccessResponse{data=object{contents=[]types.CrawledContentResponse,total=int64}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /data-sources/{id}/content [get]
func GetDataSourceContent(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid data source ID"))
	}

	page := c.QueryInt("page", 1)
	limit := c.QueryInt("limit", 20)

	contents, total, err := coreDS.GetContent(c.Context(), id, page, limit)
	if err != nil {
		return response.Error(c, err)
	}

	return response.Success(c, fiber.Map{
		"contents": contents,
		"total":    total,
		"page":     page,
		"limit":    limit,
	})
}

File: backend/src/api/datasource/handlers.rs
use axum::{
    Json,
    body::Bytes,
    extract::{Path, Query, State, rejection::JsonRejection},
    http::StatusCode,
};
use uuid::Uuid;

use crate::api::{auth::AuthenticatedAccount, response::SuccessResponse};

use super::{
    dto::{
        CrawlTriggeredResponse, CrawledContentResponse, DataSourceContentResponse,
        DataSourceCreateRequest, DataSourceDiscoveryRecommendationRequest,
        DataSourceRecommendationRequest, DataSourceRecommendationsResponse, DataSourceResponse,
        DataSourceUpdateRequest, PaginationQuery, SuccessFlag,
    },
    error::DataSourceApiError,
    state::DataSourceState,
};

type ApiResult<T> = Result<Json<SuccessResponse<T>>, DataSourceApiError>;

#[utoipa::path(
    get,
    path = "/data-sources",
    params(PaginationQuery),
    responses(
        (status = 200, body = SuccessResponse<Vec<DataSourceResponse>>),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "data-sources",
    operation_id = "listDataSources"
)]
pub async fn list_data_sources(
    authenticated: AuthenticatedAccount,
    State(state): State<DataSourceState>,
    Query(_query): Query<PaginationQuery>,
) -> ApiResult<Vec<DataSourceResponse>> {
    let account_id = authenticated.into_inner();
    let organization_id = state.organization_id(account_id).await;
    let values = match organization_id {
        Some(id) => state.service().list(id).await?,
        None => {
            state
                .service()
                .list_by_user_id(account_id.into_inner())
                .await?
        }
    };
    Ok(Json(SuccessResponse::new(
        values.into_iter().map(Into::into).collect(),
    )))
}

#[utoipa::path(
    post,
    path = "/data-sources",
    request_body = DataSourceCreateRequest,
    responses(
        (status = 201, body = SuccessResponse<DataSourceResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 409, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "data-sources",
    operation_id = "createDataSource"
)]
pub async fn create_data_source(
    authenticated: AuthenticatedAccount,
    State(state): State<DataSourceState>,
    body: Result<Json<DataSourceCreateRequest>, JsonRejection>,
) -> Result<(StatusCode, Json<SuccessResponse<DataSourceResponse>>), DataSourceApiError> {
    let Json(request) = body.map_err(invalid_request_body)?;
    let account_id = authenticated.into_inner();
    let organization_id = state.organization_id(account_id).await;
    let user_id = organization_id.is_none().then(|| account_id.into_inner());
    let value = state
        .service()
        .create(
            organization_id,
            user_id,
            request.validate().map_err(DataSourceApiError::validation)?,
        )
        .await?;
    Ok((
        StatusCode::CREATED,
        Json(SuccessResponse::new(value.into())),
    ))
}

#[utoipa::path(
    post,
    path = "/data-sources/recommendations",
    request_body = DataSourceRecommendationRequest,
    responses(
        (status = 200, body = SuccessResponse<DataSourceRecommendationsResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 502, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "data-sources",
    operation_id = "recommendDataSources"
)]
pub async fn recommend_data_sources(
    authenticated: AuthenticatedAccount,
    State(state): State<DataSourceState>,
    body: Result<Json<DataSourceRecommendationRequest>, JsonRejection>,
) -> ApiResult<DataSourceRecommendationsResponse> {
    let Json(request) = body.map_err(invalid_request_body)?;
    let account_id = authenticated.into_inner();
    let organization_id = state.organization_id(account_id).await;
    let user_id = organization_id.is_none().then(|| account_id.into_inner());
    let value = state
        .recommendations()
        .recommend(
            organization_id,
            user_id,
            request.validate().map_err(DataSourceApiError::validation)?,
        )
        .await?;
    Ok(Json(SuccessResponse::new(value.into())))
}

#[utoipa::path(
    post,
    path = "/data-sources/recommendations/discovery",
    request_body = Option<DataSourceDiscoveryRecommendationRequest>,
    responses(
        (status = 200, body = SuccessResponse<DataSourceRecommendationsResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 502, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "data-sources",
    operation_id = "discoverDataSources"
)]
pub async fn discover_data_sources(
    authenticated: AuthenticatedAccount,
    State(state): State<DataSourceState>,
    body: Bytes,
) -> ApiResult<DataSourceRecommendationsResponse> {
    let request = if body.is_empty() {
        DataSourceDiscoveryRecommendationRequest::default()
    } else {
        serde_json::from_slice(&body)
            .map_err(|_| crate::error::AppError::InvalidInput("Invalid request body".to_owned()))?
    };
    let account_id = authenticated.into_inner();
    let organization_id = state.organization_id(account_id).await;
    let user_id = organization_id.is_none().then(|| account_id.into_inner());
    let value = state
        .recommendations()
        .recommend_from_existing_sources(organization_id, user_id, request.into())
        .await?;
    Ok(Json(SuccessResponse::new(value.into())))
}

#[utoipa::path(
    get,
    path = "/data-sources/{id}",
    params(("id" = Uuid, Path)),
    responses(
        (status = 200, body = SuccessResponse<DataSourceResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "data-sources",
    operation_id = "getDataSource"
)]
pub async fn get_data_source(
    _authenticated: AuthenticatedAccount,
    State(state): State<DataSourceState>,
    Path(id): Path<String>,
) -> ApiResult<DataSourceResponse> {
    let id = parse_data_source_id(&id)?;
    Ok(Json(SuccessResponse::new(
        state.service().get_by_id(id).await?.into(),
    )))
}

#[utoipa::path(
    put,
    path = "/data-sources/{id}",
    params(("id" = Uuid, Path)),
    request_body = DataSourceUpdateRequest,
    responses(
        (status = 200, body = SuccessResponse<DataSourceResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 409, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "data-sources",
    operation_id = "updateDataSource"
)]
pub async fn update_data_source(
    _authenticated: AuthenticatedAccount,
    State(state): State<DataSourceState>,
    Path(id): Path<String>,
    body: Result<Json<DataSourceUpdateRequest>, JsonRejection>,
) -> ApiResult<DataSourceResponse> {
    let id = parse_data_source_id(&id)?;
    let Json(request) = body.map_err(invalid_request_body)?;
    Ok(Json(SuccessResponse::new(
        state.service().update(id, request.into()).await?.into(),
    )))
}

#[utoipa::path(
    delete,
    path = "/data-sources/{id}",
    params(("id" = Uuid, Path)),
    responses(
        (status = 200, body = SuccessResponse<SuccessFlag>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "data-sources",
    operation_id = "deleteDataSource"
)]
pub async fn delete_data_source(
    _authenticated: AuthenticatedAccount,
    State(state): State<DataSourceState>,
    Path(id): Path<String>,
) -> ApiResult<SuccessFlag> {
    let id = parse_data_source_id(&id)?;
    state.service().delete(id).await?;
    Ok(Json(SuccessResponse::new(SuccessFlag { success: true })))
}

#[utoipa::path(
    post,
    path = "/data-sources/{id}/crawl",
    params(("id" = Uuid, Path)),
    responses(
        (status = 200, body = SuccessResponse<CrawlTriggeredResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "data-sources",
    operation_id = "triggerDataSourceCrawl"
)]
pub async fn trigger_crawl(
    _authenticated: AuthenticatedAccount,
    State(state): State<DataSourceState>,
    Path(id): Path<String>,
) -> ApiResult<CrawlTriggeredResponse> {
    let id = parse_data_source_id(&id)?;
    state.service().trigger_crawl(id).await?;
    Ok(Json(SuccessResponse::new(CrawlTriggeredResponse {
        success: true,
        message: "Crawl triggered successfully",
    })))
}

#[utoipa::path(
    get,
    path = "/data-sources/{id}/content",
    params(("id" = Uuid, Path), PaginationQuery),
    responses(
        (status = 200, body = SuccessResponse<DataSourceContentResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "data-sources",
    operation_id = "getDataSourceContent"
)]
pub async fn get_data_source_content(
    _authenticated: AuthenticatedAccount,
    State(state): State<DataSourceState>,
    Path(id): Path<String>,
    Query(query): Query<PaginationQuery>,
) -> ApiResult<DataSourceContentResponse> {
    let id = parse_data_source_id(&id)?;
    let page = query.page.unwrap_or(1);
    let limit = query.limit.unwrap_or(20);
    let (contents, total) = state.service().get_content(id, page, limit).await?;
    Ok(Json(SuccessResponse::new(DataSourceContentResponse {
        contents: contents
            .into_iter()
            .map(CrawledContentResponse::from)
            .collect(),
        total,
        page,
        limit,
    })))
}

fn parse_data_source_id(value: &str) -> Result<Uuid, DataSourceApiError> {
    Uuid::parse_str(value).map_err(|_| {
        crate::error::AppError::InvalidInput("Invalid data source ID".to_owned()).into()
    })
}

fn invalid_request_body(_: JsonRejection) -> DataSourceApiError {
    crate::error::AppError::InvalidInput("Invalid request body".to_owned()).into()
}
